File: preprocess_all_data.m
function preprocess_all_data()
    clear; clc; close all;
    
    % ----------- 1) Define the folder pairs (input -> output) -----------
    folderPairs = {
        'ADHD_part1',    'ADHD_part1_preprocessed';
        'ADHD_part2',    'ADHD_part2_preprocessed';
        'Control_part1', 'Control_part1_preprocessed';
        'Control_part2', 'Control_part2_preprocessed'
    };
    
    % ----------- 2) Define Preprocessing Parameters -----------
    fs = 128;               % Sampling frequency
    lowCut = 1;             % Bandpass lower cutoff (Hz)
    highCut = 40;           % Bandpass upper cutoff (Hz)
    filterOrder = 4;        % Order of the Butterworth filter
    artifactThreshold = 1000;  % Simple amplitude threshold (µV)
    
    % Some filters require >= 3*(2*filterOrder) samples to run filtfilt
    minLengthForFilter = 3 * (2 * filterOrder);
    
    % Design the Butterworth filter once
    [b, a] = butter(filterOrder, [lowCut, highCut] / (fs/2), 'bandpass');
    
    % ----------- 3) Loop Over All Folder Pairs -----------
    for iPair = 1:size(folderPairs,1)
        
        inputDir  = folderPairs{iPair, 1};
        outputDir = folderPairs{iPair, 2};
        
        % Create output directory if needed
        if ~exist(outputDir, 'dir')
            mkdir(outputDir);
        end
        
        % List all .mat files in the input directory
        fileList = dir(fullfile(inputDir, '*.mat'));
        fprintf('\n=== Processing folder: %s ===\n', inputDir);
        
        if isempty(fileList)
            warning('No .mat files found in: %s', inputDir);
            continue;
        end
        
        % 3A) Process Each File
        for iFile = 1:length(fileList)
            
            inFile = fileList(iFile).name;
            inPath = fullfile(fileList(iFile).folder, inFile);
            
            fprintf('\n  -> Processing file [%d/%d]: %s\n', ...
                iFile, length(fileList), inFile);
            
            % ----------- Load Data -----------
            dataStruct = load(inPath);
            fn = fieldnames(dataStruct);
            rawData = dataStruct.(fn{1});  % e.g. dataStruct.v1p
            
            % Ensure channels x time orientation
            [rows, cols] = size(rawData);
            if rows < cols
                % If we suspect it's time x channels, transpose so
                % it becomes channels x time
                rawData = rawData';
                [rows, cols] = size(rawData);
            end
            % Now rawData is (channels x time)
            channels   = rows;
            timePoints = cols;
            
            % ----------- Filter (if data is long enough) -----------
            if timePoints < minLengthForFilter
                % If the data is too short to safely run filtfilt,
                % we skip the filter step
                fprintf('    * WARNING: Data length (%d) < %d required. Skipping filter.\n', ...
                    timePoints, minLengthForFilter);
                filteredData = rawData;
            else
                % Apply filtfilt channel-by-channel
                filteredData = zeros(size(rawData));
                for ch = 1:channels
                    filteredData(ch,:) = filtfilt(b, a, rawData(ch,:));
                end
            end
            
            % ----------- Artifact Rejection by Threshold -----------
            artifactMask = abs(filteredData) > artifactThreshold;
            cleanedData = filteredData;
            cleanedData(artifactMask) = NaN;
            
            % ----------- Channel-wise Normalization -----------
            normalizedData = zeros(size(cleanedData));
            for ch = 1:channels
                chanData = cleanedData(ch,:);
                goodIdx  = ~isnan(chanData);
                
                if any(goodIdx)
                    mu       = mean(chanData(goodIdx));
                    sigmaVal = std(chanData(goodIdx));
                    if sigmaVal < 1e-10
                        sigmaVal = 1;  % avoid /0
                    end
                    normalizedData(ch,goodIdx) = (chanData(goodIdx) - mu) / sigmaVal;
                else
                    % If entire channel is NaN, remain zeros or NaNs
                    normalizedData(ch,:) = NaN;
                end
            end
            
            preprocessedData = normalizedData;
            
            % ----------- Save Output -----------
            [~, baseName] = fileparts(inFile);
            outFileName = [baseName, '_preprocessed.mat'];
            outPath = fullfile(outputDir, outFileName);
            
            save(outPath, 'preprocessedData', 'fs');
            fprintf('    -> Saved: %s\n', outFileName);
        end
        
        fprintf('=== Finished folder: %s ===\n', inputDir);
    end
    
    fprintf('\n*** All folders processed! ***\n');
end

File: preprocess_all_data.py
#!/usr/bin/env python3

import os
import numpy as np
from scipy.io import loadmat, savemat
from scipy.signal import butter, filtfilt

def preprocess_all_data():
    folder_pairs = [
        ('../ADHD_data/ADHD_part1',    '../ADHD_data/ADHD_part1_preprocessed'),
        ('../ADHD_data/ADHD_part2',    '../ADHD_data/ADHD_part2_preprocessed'),
        ('../ADHD_data/Control_part1', '../ADHD_data/Control_part1_preprocessed'),
        ('../ADHD_data/Control_part2', '../ADHD_data/Control_part2_preprocessed')
    ]
    
    fs = 128
    low_cut = 1.0
    high_cut = 40.0
    filter_order = 4
    artifact_threshold = 1000.0
    min_length_for_filter = 3 * (2 * filter_order)

    # Design Butterworth (1-40 Hz)
    b, a = butter(filter_order, [low_cut/(fs/2), high_cut/(fs/2)], btype='band')

    for input_dir, output_dir in folder_pairs:
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
        
        print(f"\n=== Processing folder: {input_dir} ===")
        
        mat_files = [f for f in os.listdir(input_dir) if f.endswith('.mat')]
        
        if not mat_files:
            print(f"  * WARNING: No .mat files found in: {input_dir}")
            continue
        
        for i, mat_file in enumerate(mat_files, start=1):
            in_path = os.path.join(input_dir, mat_file)
            print(f"\n -> Processing file [{i}/{len(mat_files)}]: {mat_file}")
            
            data_struct = loadmat(in_path)
            var_names = [k for k in data_struct.keys() if not k.startswith('__')]
            if not var_names:
                print(f"   * ERROR: No data variable found in {mat_file}")
                continue
            
            raw_data = data_struct[var_names[0]]
            if not isinstance(raw_data, np.ndarray) or raw_data.ndim != 2:
                print(f"   * ERROR: {var_names[0]} is not a 2D array.")
                continue
            
            rows, cols = raw_data.shape
            # FIXED: Transpose if ROWS > COLS (time,channels => channels,time)
            if rows > cols:
                raw_data = raw_data.T
                rows, cols = raw_data.shape
            
            channels = rows
            time_points = cols
            
            # Filter if enough samples
            if time_points < min_length_for_filter:
                print(f"    * WARNING: Data length ({time_points}) < {min_length_for_filter}. Skipping filter.")
                filtered_data = raw_data.copy()
            else:
                filtered_data = np.zeros_like(raw_data)
                for ch in range(channels):
                    filtered_data[ch, :] = filtfilt(b, a, raw_data[ch, :])
            
            # Artifact threshold
            artifact_mask = np.abs(filtered_data) > artifact_threshold
            cleaned_data = filtered_data.copy().astype(float)
            cleaned_data[artifact_mask] = np.nan
            
            # Z-score normalization per channel (ignore NaN)
            normalized_data = np.zeros_like(cleaned_data)
            for ch in range(channels):
                chan_data = cleaned_data[ch, :]
                good_idx = ~np.isnan(chan_data)
                if np.any(good_idx):
                    mu = np.mean(chan_data[good_idx])
                    sigma_val = np.std(chan_data[good_idx])
                    if sigma_val < 1e-10:
                        sigma_val = 1.0
                    normalized_data[ch, good_idx] = (chan_data[good_idx] - mu) / sigma_val
                else:
                    normalized_data[ch, :] = np.nan
            
            preprocessed_data = normalized_data
            base_name, _ = os.path.splitext(mat_file)
            out_file_name = base_name + '_preprocessed.mat'
            out_path = os.path.join(output_dir, out_file_name)
            
            savemat(out_path, {
                'preprocessedData': preprocessed_data,
                'fs': fs
            })
            print(f"    -> Saved: {out_file_name}")
        
        print(f"=== Finished folder: {input_dir} ===")

    print("\n*** All folders processed! ***\n")


if __name__ == "__main__":
    preprocess_all_data()
